org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/text/correction/SourceAssistProcessor.java

```diff
@@ -138,8 +140,33 @@ public SourceAssistProcessor(PreferenceManager preferenceManager) {
 
 		// Generate hashCode() and equals()
 		if (supportsHashCodeEquals(context, type, monitor)) {
-			Optional<Either<Command, CodeAction>> hashCodeEquals = getHashCodeEqualsAction(params);
-			addSourceActionCommand($, params.getContext(), hashCodeEquals);
+			// Generate QuickAssist
+			Optional<Either<Command, CodeAction>> quickAssistHashCodeEquals = Optional.empty();
+			ASTNode node = context.getCoveredNode();
+			if (node == null) {
+				node = context.getCoveringNode();
+			}
+			if (node instanceof SimpleName) {
+				ASTNode parent = node.getParent();
+				if (parent instanceof TypeDeclaration) {
+					quickAssistHashCodeEquals = getHashCodeEqualsAction(params, JavaCodeActionKind.QUICK_ASSIST);
+					addSourceActionCommand($, params.getContext(), quickAssistHashCodeEquals);
+				}
+			}
+
+			// Generate Source Action
+			Optional<Either<Command, CodeAction>> sourceActionHashCodeEquals = Optional.empty();
+			if (quickAssistHashCodeEquals.isEmpty()) {
+				sourceActionHashCodeEquals = getHashCodeEqualsAction(params, JavaCodeActionKind.SOURCE_GENERATE_HASHCODE_EQUALS);
+			} else {
+				sourceActionHashCodeEquals = copySourceActionCommand(quickAssistHashCodeEquals);
+				Either<Command, CodeAction> eitherValue = sourceActionHashCodeEquals.get();
+				if (eitherValue.isRight()) {
+					eitherValue.getRight().setKind(JavaCodeActionKind.SOURCE_GENERATE_HASHCODE_EQUALS);
+				}
+			}
+			addSourceActionCommand($, params.getContext(), sourceActionHashCodeEquals);

```

Comment:
Since `getHashCodeEqualsAction()` is only used to build a CodeAction instance, there is no additional calculation logic in it. I would want to use it to directly recreate an instance for `sourceActionHashCodeEquals` instead of copying from `quickAssistHashCodeEquals`. This way the code would be shorter and more readable.
```java
Optional<Either<Command, CodeAction>> sourceActionHashCodeEquals = getHashCodeEqualsAction(params, JavaCodeActionKind.SOURCE_GENERATE_HASHCODE_EQUALS);
```

Reply:
OK. I'll separate these processes then.